### ISSUE TITLE
Discussion: Install netdata on all servers

### ISSUE THREAD

**Issue by blarghmatey** [mitodl/salt-ops#27]
NetData is an open source tool that provides realtime metrics on a per-server basis. This is potentially useful for ad-hoc debugging/investigation purposes. https://github.com/firehol/netdata

**Comment by markbreedlove** [comment]
I'm checking how the installation goes on an instance that I've set up as a candidate to replace the OCW CMS 2 server.

**Comment by markbreedlove** [comment]
Correction: I've installed it on `ocw-qa-cms-2`, instead.

**Comment by markbreedlove** [comment]
I have this installed on three of our OCW QA servers. Here are some observations.

The easiest-to-read netdata documentation is on [their wiki](https://docs.netdata.cloud/).

Their recommended installation is not package-based. They have Debian packages, but they're considered "beta."  The recommended installation employs a script to compile and install netdata, and they alternatively have a prebuilt binary installer that is a second choice. The installer scripts set the system up to upgrade netdata automatically every day from their nightly stable release. Once the installation is done, therefore, there's theoretically no need for us to deal with upgrades. The installation creates an uninstallation shell script, in case we ever need to remove netdata.

Since netdata runs on each instance, it's useless and its historical data are inaccessible if the instance becomes unavailable, due to networking problems, out-of-memory issues, etc. I have not evaluated Netdata Cloud or other installation options besides the basic one-installation-per-instance kind.

It appears that their dashboard can be cusomized, but this seems to involve working on some Javascript files and writing code. You can't create a custom dashboard in the U.I. that groups particular charts that you want to see next to each other.

For Nginx stats, we'd have to compile `nginx` with the `ngx_http_stub_status_module` module and enable it in a `location /stub_status`.

On the other hand, their Elasticsearch module doesn't appear to have any dependencies besides the existence of an  Elasticsearch installation.

I haven't evaluated every module, but there are others for a variety of services that we use.

I can create a Salt state to install Netdata, and another one to manage its configuration files. Out of the box I had to make one change to the default configuration, to get it to save data to its round-robin database on disk, instead of persisting the stats only in memory. Out of the box, it only saves one hour of statistics, in memory.

Overall, I think the reports could be very useful for diagnosing issues, because of their granularity, and because of various conveniences provided by the U.I. It works quite well with the basic installer.

**Comment by markbreedlove** [comment]
This issue is on hold for now.

**Comment by blarghmatey** [comment]
After another look they do provide a static build of netdata which we can install without requiring installation of `-dev` dependencies and compiling the code on the instances. Given this information I think it is worth continuing in this effort to get more granular metrics for the purpose of debugging.

https://docs.netdata.cloud/packaging/installer/#linux-64bit-pre-built-static-binary

I think it's worth decomposing the shell script that they reference into a salt formula.

**Comment by markbreedlove** [comment]
Initial PR for `netdata-formula`: https://github.com/mitodl/netdata-formula/pull/1

**Comment by markbreedlove** [comment]
https://github.com/mitodl/netdata-formula/pull/2 upgrades to Python3, after some commits recently to fix things.

**Comment by markbreedlove** [comment]
That PR was merged on Monday and the state was run on our QA elasticsearch nodes. It worked, and reports look good. I've added a security group for the relevant VPN and will do so for other VPNs as I proceed to execute the state on more of our instances.

**Comment by markbreedlove** [comment]
The security group will allow access for ODL Engineering in the office or on the VPN.

**Comment by markbreedlove** [comment]
I removed that security group yesterday and added the necessary rules to allow access via SSH tunneling, using the existing "master-ssh" security groups.
